crates/ide-diagnostics/src/handlers/external_lints.rs

```diff
@@ -0,0 +1,156 @@
+        ty::Mutability,
+        BodyId, Crate, CrateId, Span, Symbol,
+    },
+    context::AstContext,
+};
+use once_cell::unsync::OnceCell;
+use syntax::{
+    ast::{HasModuleItem, HasVisibility, Item},
+    SourceFile, SyntaxNode, TextRange, TextSize, AstNode,
+};
+
+use crate::{Diagnostic, DiagnosticsConfig};
+
+struct RADriver<'a> {
+    diags: RefCell<&'a mut Vec<Diagnostic>>,
+    id_map: RefCell<&'a mut Vec<SyntaxNode>>,
+    ast_context: OnceCell<&'a AstContext<'a>>,
+}
+
+fn leak<'a, T: 'a>(x: T) -> &'a T {
+    Box::leak(Box::new(x))
+}
+
+impl<'ast> DriverContext<'ast> for RADriver<'ast> {
+    fn emit_lint(
+        &self,
+        lint: &'static linter_api::lint::Lint,
+        msg: &str,
+        span: &linter_api::ast::Span<'ast>,
+    ) {
+        self.diags.borrow_mut().push(Diagnostic::new(
+            lint.name,
+            msg,
+            TextRange::new(TextSize::from(span.start() as u32), TextSize::from(span.end() as u32)),
+        ));
+    }
+
+    fn get_span(
+        &'ast self,
+        owner: &linter_api::ast::SpanOwner,
+    ) -> &'ast linter_api::ast::Span<'ast> {
+        match owner {
+            linter_api::ast::SpanOwner::Item(id) => {
+                let span = self.id_map.borrow().get(id.get_data().1 as usize).unwrap().text_range();
+                leak(Span::new(
+                    self.ast_context.get().unwrap(),
+                    linter_api::ast::SpanSource::File(leak(PathBuf::try_from("/").unwrap())),
+                    span.start().into(),
+                    span.end().into(),
+                ))
+            }
+            linter_api::ast::SpanOwner::Body(_) => todo!(),
+            linter_api::ast::SpanOwner::SpecificSpan(_) => todo!(),
+        }
+    }
+
+    fn span_snippet(&self, _: &linter_api::ast::Span<'_>) -> Option<&'ast str> {
+        todo!()
+    }
+}
+
+pub(crate) fn external_lints(
+    acc: &mut Vec<Diagnostic>,
+    _: FileId,
+    node: &SourceFile,
+    config: &DiagnosticsConfig,
+) {
+    std::env::set_var("LINTER_LINT_CRATES", &config.external_lints);
+    let mut x = linter_adapter::Adapter::new_from_env();
+    let mut id_map = Vec::new();
+    let driver = RADriver {
+        diags: RefCell::new(acc),
+        id_map: RefCell::new(&mut id_map),
+        ast_context: OnceCell::new(),
+    };
+    let cw = DriverContextWrapper::new(&driver);
+    let c = cw.create_driver_callback();
+    let a = AstContext::new(&c);
+    driver.ast_context.set(&a).unwrap();
+    let crate_id = CrateId::new(0);
+    let items: Vec<_> =
+        node.items().filter_map(|item| convert_item(item, &driver, crate_id)).collect();
+    x.process_krate(&a, &Crate::new(crate_id, &items));
+}
+
+fn convert_item<'ast>(
+    item: Item,
+    driver: &'ast RADriver<'ast>,
+    crate_id: CrateId,
+) -> Option<ItemType<'ast>> {
+    let cx = &driver.ast_context.get().unwrap();
+    let item_id = ItemId::new(crate_id, driver.id_map.borrow().len() as u32);
```

Comment:
At ast level r-a doesn't have id for items, so I made it in this way. Is it possible / better to use `&'ast Item` instead of `ItemId`?

Reply:
What does ra use to identify items then?

I changed the `ItemId` to just be a `u64` which can be used by the driver however they want. In theory, a pointer address can also be stored in it. The problem with this solution, is that it requires the item to already be created, which goes a bit against lazy evaluation, which was also wished for if I'm correct.